src/dctest/expressions.cljs

```diff
@@ -78,14 +90,65 @@ END_PROP_BRACK   ::= WHITESPACE* ']' WHITESPACE*
 Identifier       ::= IDENTIFIER_START IDENTIFIER_PART*
 IDENTIFIER_START ::= [a-zA-Z] | '$' | '_'
 IDENTIFIER_PART  ::= [a-zA-Z0-9] | '$' | '_'
+
+PrintableChar ::= #x0009 | #x000A | #x000D | [#x0020-#xD7FF] | [#xE000-#xFFFD] | [#x10000-#x10FFFF]
+WHITESPACE    ::= [#x20#x09#x0A#x0D]+   /* Space | Tab | \\n | \\r */
+
+/* Used for error reporting, not evaluated directly */
+ExpectedInterpolation ::= InterpolatedExpression PrintableChar*
   ")
 
 (def parser
   (ebnf/Grammars.W3C.Parser. grammar))
 
+(defn semantic-errors [ast]
+  (let [{:keys [children text type]} ast]
+    (case type
+      "InterpolatedText" (let [positions (interp-positions text)
+                               expressions (filter #(= "InterpolatedExpression" (:type %)) children)]
+                           (when-not (= (count positions)
+                                        (count expressions))
+                             ;; Provide location of unparseable/unclosed expressions
+                             (mapcat (fn [pos]
+                                       (when-not (-> (subs text pos)
+                                                     (read-ast "ExpectedInterpolation")
+                                                     :children
+                                                     first)
+                                         [{:message (str "Invalid Expression at position " pos)}]))
+                                     positions)))
+      "ContextName" (when-not (contains? supported-contexts text)
+                      [{:message (str "ReferenceError: " text " is not supported")}])
+      "FunctionCall" (let [[func & args] children
+                           func-name (:text func)
+                           func (get stdlib func-name)]
+                       (cond
+                         (not func)
+                         [{:message (str "ReferenceError: " func-name " is not supported")}]
+
+                         (not= (count args) (:arity func))
+                         [{:message (str "ArityError: incorrect number of arguments to " func-name)}]
+
+                         :else nil))
+      ;; else
+      nil)))
+
+(defn check-ast [ast]
+  (let [errors (or (semantic-errors ast) [])]
+    (-> ast
+        (update :children #(mapv check-ast %))
```

Comment:
You're right, thanks! I've replaced 'check-ast and 'flatten-errors with versions that use zippers.  The two function share a bit of structure for zipper navigation, but I think it might be clearer if it's duplicated instead of extracted.  Let me know if you agree or if you prefer another extracted function here.